PSYKHANOOL/PSYKHANOOL.h

```diff
@@ -0,0 +1,664 @@
+//
+// * OrderedDictionary<T> <=> std::map<T_KEY, T>
+//   Empty(), Size(), operator[](key), Erase(key) [, iteration, {lower/upper}_bound].
+//   `T_KEY` is either the type of `T.key` or of `T.get_key()`.
+//
+// * LightweightMatrix<T> <=> { T_ROW, T_COL } -> T, three `std::map<>`-s.
+//   Empty(), Size(), Rows()/Cols(), Add(cell), Delete(row, col) [, iteration, {lower/upper}_bound].
+//   `T_ROW` and `T_COL` are either the type of `T.row` / `T.col`, or of `T.get_row()` / `T.get_col()`.
+//
+// ...
+//
+// All Current-friendly types support persistence.
+//
+// Only allow default constructors for containers.
+
+#ifndef PSYKHANOOL_H
+#define PSYKHANOOL_H
+
+#include <fstream>
+#include <vector>
+#include <map>
+#include <utility>
+
+#include "sfinae.h"
+
+#include "../Bricks/time/chrono.h"
+#include "../Bricks/strings/strings.h"
+#include "../Bricks/cerealize/cerealize.h"
+
+namespace PSYKHANOOL {
+
+struct NoValueException : std::exception {};
+typedef const NoValueException& NoValue;
+
+struct CannotPopBackFromEmptyVectorException : std::exception {};
+typedef const CannotPopBackFromEmptyVectorException& CannotPopBackFromEmptyVector;
+
+template <typename T>
+class ImmutableOptional final {
+ public:
+  ImmutableOptional() = delete;
+  ImmutableOptional(const T* object) : optional_object_(object) {}
+  ImmutableOptional(std::unique_ptr<T>&& rhs)
+      : owned_optional_object_(std::move(rhs)), optional_object_(owned_optional_object_.get()) {}
+  bool Exists() const { return optional_object_ != nullptr; }
+  const T& Value() const {
+    if (optional_object_ != nullptr) {
+      return *optional_object_;
+    } else {
+      throw NoValueException();  // TBD: Derived detailed exception type.
+    }
+  }
+
+ private:
+  std::unique_ptr<T> owned_optional_object_;
+  const T* optional_object_;
+};
+
+template <typename T>
+class MutableOptional final {
+ public:
+  MutableOptional() = delete;
+  MutableOptional(T* object) : optional_object_(object) {}
+  bool Exists() const { return optional_object_ != nullptr; }
+  T& Value() {
+    if (optional_object_ != nullptr) {
+      return *optional_object_;
+    } else {
+      throw NoValueException();  // TBD: Derived detailed exception type.
+    }
+  }
+
+ private:
+  T* optional_object_;
+};
+
+template <typename T>
+class VectorStorage {
+ public:
+  std::vector<T> vector_;
+};
+
+template <typename T, typename PERSISTER>
+class VectorAPI : protected VectorStorage<T> {
+ public:
+  explicit VectorAPI(PERSISTER* persister) : persister_(*persister) {}
+
+  bool Empty() const { return VectorStorage<T>::vector_.empty(); }
+  size_t Size() const { return VectorStorage<T>::vector_.size(); }
+
+  typename PERSISTER::MutableOptionalType operator[](size_t index) {
+    if (index < VectorStorage<T>::vector_.size()) {
+      return typename PERSISTER::MutableOptionalType(&VectorStorage<T>::vector_[index]);
+    } else {
+      return typename PERSISTER::MutableOptionalType(nullptr);
+    }
+  }
```

Comment:
Do I understand correctly that user can modify vector elements and these changes will not be persisted?